src/multi/uvarint.cpp

```diff
@@ -91,12 +77,24 @@ namespace libp2p::multi {
   }
 
   size_t UVarint::calculateSize(gsl::span<const uint8_t> varint_bytes) {
-    size_t s = 0;
-
-    while ((varint_bytes[s] & 0x80u) != 0) {
-      s++;
+    size_t size = 0;
+    size_t shift = 0;
+    constexpr size_t capacity = sizeof(uint64_t) * 8;
+    bool last_byte_found = false;
+    for (const auto &byte : varint_bytes) {
+      ++size;
+      uint64_t slice = byte & 0x7f;
+      if (shift >= capacity || slice << shift >> shift != slice) {
+        size = 0;
+        break;
+      }
+      if ((byte & 0x80) == 0) {
+        last_byte_found = true;
+        break;
+      }
+      shift += 7;
     }
-    return s + 1;
+    return last_byte_found ? size : 0;
```

Comment:
To check whether a raw bytes sequence is a valid UVarint-encoded sequence we must check not only the length of the bytes, but also the contents of the last byte to prevent overflow